Add blog post for purrr 0.3.0 (#269)
tidyverse/tidyverse.org: content/articles/2019-01-purrr-0-3-0.Rmd

---
title: 'purrr 0.3.0'
author: Lionel Henry
date: '2019-01-25'
slug: purrr-0-3-0
description: >
  purrr 0.3.0 is now on CRAN.
categories:
  - package
photo:
  url: https://unsplash.com/photos/NodtnCsLdTE
  author: Mikhail Vasilyev
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.width = 7,
  fig.align = 'center',
  fig.asp = 0.618, # 1 / phi
  out.width = "700px"
)

library("purrr")
```

We're excited to announce the release of purrr 0.3.0! This release polishes the consistency of the interface, introduces new `pluck()`-based functions, many improvements to mappers, and a set of tools to limit the rate at which a function is called. Find a detailed account of the changes in the [NEWS](https://github.com/tidyverse/purrr/blob/master/NEWS.md#purrr-030) file.

> 

Reviewer: I think we should always reveal all the necessary code, having watched some people live code based on quick Googling. So that means always making this explicit:

  * Install the latest version via `install.packages("purrr")`.
  * Attach (or load?) purrr via `library(purrr)`.

Reviewer 2: Also needs a brief overview of the rest of the documentation. It currently feels like a grab bag of different features; that's fine, but we should prepare the user for it.

Author: Instead of attach/load: "Start working with purrr"?

## Pluck assignment

> 

Reviewer: Maybe include a quick reminder of what `pluck()` is for? Quote or paraphrase from the docs:

> `pluck()` implements a generalised form of `[[` that allow you to index deeply and flexibly into data structures.

The new functions `pluck<-`, `assign_in()` and `modify_in()` are assignment variants of `pluck()`. Let's create a nested data structure:

```{r}
x <- list(foo = list(1, 2), bar = list(3, 4))
str(x)
```

> This sort of repeated structures is the kind of data where `pluck()` shines:

Reviewer: ```suggestion
This sort of repeated structure is the kind of data where `pluck()` shines:
```

```{r}
pluck(x, "foo", 2)

pluck(x, "bar", 1)
```

You can now use the same syntax to modify the data:

```{r}
pluck(x, "foo", 2) <- 100
str(x)
```

> `pluck<-` also has a functional form that won't modify objects in your environment, but instead returns the modified data:

Reviewer: ```suggestion
`pluck<-` also has a functional form that does not modify objects in your environment, but instead returns a modified copy:
```

```{r}
out <- assign_in(x, list("foo", 2), 2000)

# The object is still the same as before
str(x)

# The modified data is in `out`
str(out)
```

Finally, `modify_in()` is a variant of `modify()` that only changes the pluck location with the result of applying a function:

```{r}
out <- modify_in(x, list("foo", 2), as.character)
str(out)
```

> assign_in() and modify_in()

Reviewer: Seems to be leftover from writing process?

```{r}
f <- function(...) print(Sys.time())

walk(1:3, f)

walk(1:3, slowly(f))
```

`slowly()` uses a constant rate by default while `insistently()` uses a backoff rate. The rate limiting can be configured with optional jitter via `rate_backoff()` and `rate_delay()`, which implement exponential backoff rate and constant rate respectively.

```{r}
walk(1:3, slowly(f, rate_backoff(max_times = Inf)))
```

## `map_if()`... or else?

If you like using `map_if()`, perhaps you'll find the new `.else` argument useful. `.else` is a function applied to elements for which the predicate is `FALSE`:

```{r}
map_if(iris, is.numeric, mean, .else = nlevels)
```

## Strictness

### Stricter predicate checking

purrr now checks the results of your predicate functions, which must now consistently return `TRUE` or `FALSE`. We no longer offer support for `NA` or for boolish numeric values (R normally interprets 0 as `FALSE` and all other values as `TRUE`). The purpose of this change is to detect errors earlier with a more relevant error messages.

```{r, error = TRUE}
keep(c(1, NA, 3), ~ . %% 2 == 0)
```

> ### Stricter pluck()

Reviewer 2: Move up with `pluck()`?

> Thanks to Daniel Barnett (@daniel-barnett on Github), `pluck()` now has a stricter cousin `chuck()`. Whereas `pluck()` is very permissive regarding non-existing locations and returns `NULL` in these cases, `chuck()` fails consistently with informative messages:

Reviewer 2: (i.e. it "chucks" an error message)

```{r, error = TRUE}
pluck(list(1), "foo")

chuck(list(1), "foo")
```

## New `map_at()` features

Colin Fay (@ColinFay) has added support for tidyselect expressions to `map_at()` and other `_at` mappers. This brings the interface of these functions closer to scoped functions from the dplyr package, such as `dplyr::mutate_at()`. Note that `vars()` is currently not reexported from purrr, so you need to use `dplyr::vars()` or `ggplot2::vars()` for the time being.

```{r}
suppressMessages(library("dplyr"))

x <- list(
  foo = 1:5,
  bar = 6:10,
  baz = 11:15
)

map_at(x, vars(starts_with("b")), mean)
```

`map_at()` now also supports negative selections:

```{r}
map_at(x, -2, `*`, 1000)
```

## Consistency of behaviour

A lot of effort went towards more consistent behaviour in purrr. An important improvement is that `modify()` is now a wrapper around `[[<-` rather than `[<-`. This should make it directly compatible with a larger set of vector classes.

> Thanks to the work of Mikko Marttila (@mikmart), `pmap()` and `pwalk()` do a better job of preserving S3 classes. `pluck()` also work better with S3 objects. In the next version of purrr, we plan to fix once and for all this sort of issues by integrating the vctrs package to deal with vector operations in a truly generic way.

Reviewer: ```suggestion
Thanks to the work of Mikko Marttila (@mikmart), `pmap()` and `pwalk()` do a better job of preserving S3 classes. `pluck()` also works better with S3 objects. In the next version of purrr, we plan to use the in-development vctrs package to provide more principled and predictable vector operations.
```

To say "fix once and for all" just sounds like famous last words 😁

Other small improvements in consistency of behaviour are:

* `modify_depth()` now operates at the atomic level as well.
* `map()` and `modify()` now work with calls and pairlists.

## Consistency of the interface

### Direction of application

The direction of application is now specified the same way across purrr functions. `reduce()`, `compose()` and `detect()` now have a `.dir` parameter that can take the value `"forward"` or `"backward"`. This terminology should be less ambiguous than "left" and "right":

```{r, eval = FALSE}
reduce(1:4, `-`, .dir = "backward")

compose(foo, bar, .dir = "forward")

detect(1:5, ~ . %% 2 == 0, .dir = "backward")
```

> Note that the backward version of `reduce()` (called right-reduce in the literature) applies the reduced function in a slightly different way that is more consistent with how this operation

Reviewer: something's wrong with "with how this operation"
